LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.NetworkServer
+{
+    using LoRaWan.NetworkServer.BasicsStation;
+    using Microsoft.Extensions.Caching.Memory;
+    using Microsoft.Extensions.Logging;
+    using System;
+    using System.Security.Cryptography;
+    using System.Text;
+
+    public sealed class ConcentratorDeduplication : IConcentratorDeduplication, IDisposable
+    {
+        internal readonly MemoryCache Cache;
+
+        private readonly ILogger<IConcentratorDeduplication> Logger;
+        private readonly TimeSpan CacheEntryExpiration;
+
+        public ConcentratorDeduplication(
+            ILogger<IConcentratorDeduplication> logger,
+            int cacheEntryExpirationInMilliSeconds = 60_000)
+        {
+            this.Cache = new MemoryCache(new MemoryCacheOptions());
+            this.Logger = logger;
+            this.CacheEntryExpiration = TimeSpan.FromMilliseconds(cacheEntryExpirationInMilliSeconds);
+        }
+
+        public bool ShouldDrop(UpstreamDataFrame updf, StationEui stationEui)
+        {
+            if (updf == null) throw new ArgumentNullException(nameof(updf));
+
+            var key = CreateCacheKey(updf);
+            StationEui previousStation;
+
+            lock (this.Cache)
```

Comment:
`MemoryCache` is thread-safe so seems odd to me to have to lock on it.

Reply:
I tagged you in the other thread from Mik to keep the discussion in one place. I believe the [specific implementation](https://source.dot.net/#Microsoft.Extensions.Caching.Memory/MemoryCache.cs,99099576db12fef5) we use is not thread-safe.